File: implementations/rust/src/number.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum Number {
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
}

impl Number {
    pub const ARRAY_ALL: [Self; 9] = [
        Self::One,
        Self::Two,
        Self::Three,
        Self::Four,
        Self::Five,
        Self::Six,
        Self::Seven,
        Self::Eight,
        Self::Nine,
    ];

    pub fn to_u8(&self) -> u8 {
        match self {
            Self::One => 1,
            Self::Two => 2,
            Self::Three => 3,
            Self::Four => 4,
            Self::Five => 5,
            Self::Six => 6,
            Self::Seven => 7,
            Self::Eight => 8,
            Self::Nine => 9,
        }
    }

    pub fn from_u8(int: u8) -> Option<Self> {
        match int {
            1 => Some(Self::One),
            2 => Some(Self::Two),
            3 => Some(Self::Three),
            4 => Some(Self::Four),
            5 => Some(Self::Five),
            6 => Some(Self::Six),
            7 => Some(Self::Seven),
            8 => Some(Self::Eight),
            9 => Some(Self::Nine),
            _ => None,
        }
    }

    pub fn from_u8_normalize(int: u8) -> Self {
        match Self::from_u8(normalize(int)) {
            Some(num) => num,
            None => panic!("Invalid number: {}", int),
        }
    }

    pub fn from_char(char: char) -> Option<Self> {
        match char {
            '1' => Some(Self::One),
            '2' => Some(Self::Two),
            '3' => Some(Self::Three),
            '4' => Some(Self::Four),
            '5' => Some(Self::Five),
            '6' => Some(Self::Six),
            '7' => Some(Self::Seven),
            '8' => Some(Self::Eight),
            '9' => Some(Self::Nine),
            _ => None,
        }
    }

    pub fn to_char(self: &Self) -> char {
        match self {
            Self::One => '1',
            Self::Two => '2',
            Self::Three => '3',
            Self::Four => '4',
            Self::Five => '5',
            Self::Six => '6',
            Self::Seven => '7',
            Self::Eight => '8',
            Self::Nine => '9',
        }
    }
}

fn normalize(int: u8) -> u8 {
    if int < 1 {
        1
    } else if int > 9 {
        9
    } else {
        int
    }
}

File: implementations/zig/src/number.zig
const std = @import("std");

pub const Number = enum {
    const Self = @This();
    one,
    two,
    three,
    four,
    five,
    six,
    seven,
    eight,
    nine,

    pub fn toInt(self: Self) u4 {
        return switch (self) {
            .one => 1,
            .two => 2,
            .three => 3,
            .four => 4,
            .five => 5,
            .six => 6,
            .seven => 7,
            .eight => 8,
            .nine => 9,
        };
    }

    pub fn fromIntClamp(int: usize) Self {
        return switch (int) {
            0, 1 => .one,
            2 => .two,
            3 => .three,
            4 => .four,
            5 => .five,
            6 => .six,
            7 => .seven,
            8 => .eight,
            else => .nine,
        };
    }

    pub fn toChar(self: Self) u8 {
        return switch (self) {
            .one => '1',
            .two => '2',
            .three => '3',
            .four => '4',
            .five => '5',
            .six => '6',
            .seven => '7',
            .eight => '8',
            .nine => '9',
        };
    }

    pub fn fromChar(char: u8) ?Self {
        return switch (char) {
            '1' => .one,
            '2' => .two,
            '3' => .three,
            '4' => .four,
            '5' => .five,
            '6' => .six,
            '7' => .seven,
            '8' => .eight,
            '9' => .nine,
            else => null,
        };
    }
};
